readium/navigator/src/main/java/org/readium/r2/navigator/tts/PublicationSpeechSynthesizer.kt

```diff
@@ -0,0 +1,553 @@
+        }
+        return utterance
+    }
+
+    /**
+     * Loads the utterances for the next publication [Content.Element] item in the given [direction].
+     */
+    private suspend fun loadNextUtterances(direction: Direction): Boolean {
+        val content = publicationIterator?.nextIn(direction)
+            ?: return false
+
+        val nextUtterances = content
+            .tokenize()
+            .flatMap { it.utterances() }
+
+        if (nextUtterances.isEmpty()) {
+            return loadNextUtterances(direction)
+        }
+
+        utterances = CursorList(
+            list = nextUtterances,
+            startIndex = when (direction) {
+                Direction.Forward -> 0
+                Direction.Backward -> nextUtterances.size - 1
+            }
+        )
+
+        return true
+    }
+
+    /**
+     * Splits a publication [Content.Element] item into smaller chunks using the provided tokenizer.
+     *
+     * This is used to split a paragraph into sentences, for example.
+     */
+    private fun Content.Element.tokenize(): List<Content.Element> =
+        tokenizerFactory(config.value.defaultLanguage ?: publication.metadata.language)
+            .tokenize(this)
+
+    /**
+     * Splits a publication [Content.Element] item into the utterances to be spoken.
+     */
+    private fun Content.Element.utterances(): List<Utterance> {
+        fun utterance(text: String, locator: Locator, language: Language? = null): Utterance? {
+            if (!text.any { it.isLetterOrDigit() })
+                return null
+
+            return Utterance(
+                text = text,
+                locator = locator,
+                language = language
+                    // If the language is the same as the one declared globally in the publication,
+                    // we omit it. This way, the app can customize the default language used in the
+                    // configuration.
+                    ?.takeIf { it != publication.metadata.language }
+            )
+        }
+
+        return when (this) {
+            is Content.TextElement -> {
+                segments.mapNotNull { segment ->
+                    utterance(
+                        text = segment.text,
+                        locator = segment.locator,
+                        language = segment.language
+                    )
+                }
+            }
+
+            is Content.TextualElement -> {
+                listOfNotNull(
+                    text
+                        ?.takeIf { it.isNotBlank() }
+                        ?.let { utterance(text = it, locator = locator) }
+                )
+            }
+
+            else -> emptyList()
+        }
+    }
+
+    /**
+     * Cancels the previous playback-related job and starts a new one with the given suspending
+     * [block].
+     *
+     * This is used to interrupt on-going commands.
+     */
+    private fun replacePlaybackJob(block: suspend CoroutineScope.() -> Unit) {
```

Comment:
That's the solution I came up with to be able to skip utterances or interrupt a running utterance. As only one playback job is allowed at any time, when you start a new command the previous one is cancelled.